src/ExaError.lua

```diff
@@ -0,0 +1,156 @@
+-- Note that `__tostring` is the metamethod called by Lua's global `tostring` function.
+-- This allows using the error message in places where Lua expects a string.
+-- @return string representation of the error object
+function ExaError:__tostring()
+    local lines = {}
+    if self._code then
+        if self._message then
+            table.insert(lines, self._code .. ": " .. self:get_message())
+        else
+            table.insert(lines, self._code)
+        end
+    else
+        if self._message then
+            table.insert(lines, self:get_message())
+        else
+            table.insert(lines, "<Missing error message. This should not happen. Please contact the software maker.>")
+        end
+    end
+    if (self._mitigations ~= nil) and (#self._mitigations > 0) then
+        table.insert(lines, "\nMitigations:\n")
+        for _, mitigation in ipairs(self._mitigations) do
+            table.insert(lines, "* " .. expand(mitigation, self._parameters))
+        end
+    end
+    return table.concat(lines, "\n")
+end
+
+--- Concatenate an error object with another object.
+-- @return String representing the concatenation.
+function ExaError.__concat(left, right)
+    return tostring(left) .. tostring(right)
+end
+
+--- Create a new instance of an error message.
+-- @param code error code
+-- @param message error message, optionally with placeholders
+-- @param[opt={}] parameters parameter definitions used to replace the placeholders
+-- @param[opt={}] mitigations mitigations users can try to solve the error
+-- @return created object
+function ExaError:new(code, message, parameters, mitigations)
+    local instance = setmetatable({}, self)
+    instance:_init(code, message, parameters, mitigations)
+    return instance
+end
+
+function ExaError:_init(code, message, parameters, mitigations)
+    self._code = code
+    self._message = message
+    self._parameters = parameters or {}
+    self._mitigations = mitigations or {}
+end
+
+--- Add mitigations.
+-- @param ... one or more mitigation descriptions
+-- @return error message object
+function ExaError:add_mitigations(...)
+    for _, mitigation in ipairs({...}) do
+        table.insert(self._mitigations, mitigation)
+    end
+    return self
+end
+
+--- Add issue ticket mitigation
+-- This is a special kind of mitigation which you should use in case of internal software errors that should not happen.
+-- For example when a path in the code is reached that should be unreachable if the code is correct.
+-- @return error message object
+function ExaError:add_ticket_mitigation()
+    table.insert(self._mitigations,
+        "This is an internal software error. Please report it via the project's ticket tracker.")
+    return self
+end
+
+--- Get the error code.
+-- @return error code
+function ExaError:get_code()
+    return self._code
+end
+
+--- Get the error message.
+-- This method supports Lua's standard string interpolation used in `string.format`.
```

Comment:
Is that true? `string.format` uses the `%s` syntax, not `{arg}`.

Reply:
Not anymore. This is a left-over from a previous implementation that was more Lua-ish and less compliant with our other error-modules.

Reply:
Removed the sentence.